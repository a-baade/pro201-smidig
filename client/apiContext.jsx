import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";
import React from "react";

export const ApiContext = React.createContext({
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },

  async allCharities() {
    return await fetchJSON("/api/charities");
  },

  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
});