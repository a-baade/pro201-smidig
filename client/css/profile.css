.wrap{
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    min-height: 90vh;
}

.main-card{
    box-sizing: border-box;
    height: 47em;
    width: 100%;
    max-width: 47em;
    border: 3px solid #35656f;
    border-radius: 2vh;
    position: relative;
}

.color-block {
    height: 7em;
    background-color: #35656f;
    border-top-left-radius: 1vh;
    border-top-right-radius: 1vh;
    position: static;
}

.first-row{
    display: flex;
    justify-content: space-between;
    height: calc(33%);
}

.img-btn{
    display: flex;
    flex-direction: column;
    margin-top: -2.4em;
    margin-left: 4em;
}

.pfp{
    width: 9em;
    height: 9em;
    border-radius: 100%;
}

.photo-upload-btn{
    border: none;
    background-color: #c4c4c4;
    border-radius: 64px;
    margin-top: 0.5em;
}

.company-name{
    margin-right: 6em;
    margin-top: 1.5em;
}

.second-row{
    display: flex;
    justify-content: space-around;
    height: calc(33%);
    margin-top: -4em;
}

.company-contact{
    margin-left: 1em;
}

.payment-details{
    margin-right: 1em;
}

.third-row{
    display: flex;
    flex-direction: column;
    height: calc(33%);
    align-items: center;
    margin-top: -0.9em;
}

.profile-btn{
    font-family: sans-serif;
    font-style: normal;
    font-weight: 400;
    font-size: 14px;

    width: 20em;
    height: 3.4em;
    border-radius: 50px;
    border: none;
    margin-top: 1.5em;

    background: #a0cfd2;
    box-shadow: 0 0.1em 0.2em rgba(0, 0, 0, 0.25);
}