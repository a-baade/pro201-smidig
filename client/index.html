<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Meliora Impact</title>
    <link rel="icon" href="pics/icons/logo.png">
    <link rel="stylesheet" href="css/navigation.css">
    <link rel="stylesheet" href="css/style.css">
    <link rel="stylesheet" href="css/frontPage.css">
    <link rel="stylesheet" href="css/login.css">
    <link rel="stylesheet" href="css/register.css">
    <link rel="stylesheet" href="css/aboutUs.css">
    <link rel="stylesheet" href="css/signup.css">
    <link rel="stylesheet" href="css/charities.css">
    <link rel="stylesheet" href="css/charityPage.css">
    <link rel="stylesheet" href="css/donate.css">
    <link rel="stylesheet" href="css/profile.css">
    <link rel="stylesheet" href="css/contact.css">
    <link rel="stylesheet" href="css/dashboard.css">
    <link rel="stylesheet" href="css/loadingWheel.css">
    <link rel="stylesheet" href="css/footer.css">
</head>
<body>
<div id="app"></div>
</body>
<script src="index.jsx" type="module"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/react/15.1.0/react.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/react/15.1.0/react-dom.min.js"></script>
</html>