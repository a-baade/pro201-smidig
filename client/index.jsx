import React from "react";
import ReactDOM from "react-dom";
import {Application} from "./application";

ReactDOM.render(<Application />, document.getElementById("app"));