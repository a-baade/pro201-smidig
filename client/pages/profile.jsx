import React from "react";
import Edit from "../pics/icons/edit.png";
import {Link} from "react-router-dom";

export function Profile({ user }) {
  return (
    <div className={"wrap page-margin"}>
      <div className={"main-card"}>
          <div className={"color-block"}></div>
          <div className={"first-row"}>
              <div className={"img-btn"}>
                  <img src={user.google.picture} alt="Profile picture" className={"pfp"}/>
                  <button className={"photo-upload-btn"}>Upload photo</button>
              </div>
              <h2 className={"company-name"}>Dyrevernalliansen</h2>
          </div>
          <div className={"second-row"}>
              <div className={"company-contact"}>
                  <h3>company contact</h3>
                  <p>{user.google.name}</p>
                  <p>+47 95262419</p>
                  <p>{user.google.email}</p>
                  <p>ceo assistant</p>
              </div>
              <div className={"edit-icons"}>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
              </div>
              <div className={"payment-details"}>
                  <h3>payment details</h3>
                  <p>{user.google.name}</p>
                  <p>12345667890</p>
                  <p>1234 4567 4567 6789</p>
              </div>
              <div className={"edit-icons"}>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
                  <img src={Edit} className={"edit"} alt={"editIcon"}/>
              </div>
          </div>
          <div className={"third-row"}>
              <div><button className={"profile-btn"}>Update</button></div>
              <Link to={"/login/endsession"}>
                  <button className={"profile-btn"}>Sign out</button>
              </Link>
          </div>
      </div>
    </div>
  );
}