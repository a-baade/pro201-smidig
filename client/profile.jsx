import React from "react";

export function Profile({ user }) {
  return (
    <div>
      <h1>Profile</h1>
      <div>
        <div>
          <h3>Name: {user.google.name}</h3>
          <h4>Email: {user.google.email}</h4>
        </div>
        <div>
          <img src={user.google.picture} alt="Profile picture" />
        </div>
      </div>
    </div>
  );
}