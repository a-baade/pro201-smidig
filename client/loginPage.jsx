import React, { useContext, useEffect, useState } from "react";
import { Link, Route, Routes, useNavigate, useParams } from "react-router-dom";
import {ApiContext} from "./apiContext";
import {randomString} from "./lib/randomString";
import Google from "./login-pic/img.png";
import Message from "./login-pic/img_2.png";
import Invisible from "./login-pic/img_4.png";
import Lock from "./login-pic/img_3.png";
import FirstLayer from "./login-pic/login-sidebar/firstlayer.png";
import SecondLayer from "./login-pic/login-sidebar/secondlayer.png";
import TopCorner from "./login-pic/login-sidebar/topcorner.png";
import BottomCorner from "./login-pic/login-sidebar/bottomcorner.png";


export function LoginCallback({ reload, config }) {
  const { provider } = useParams();
  const [error, setError] = useState();
  const navigate = useNavigate();
  const { registerLogin } = useContext(ApiContext);
  useEffect(async () => {
    const { access_token, error, error_description, state, code } =
      Object.fromEntries(
        new URLSearchParams(window.location.hash.substring(1))
      );

    const expected_state = window.sessionStorage.getItem("expected_state");
    if (!state || expected_state !== state) {
      setError("Unexpected state");
      return;
    }

    if (error || error_description) {
      setError(`Error: ${error} (${error_description})`);
      return;
    }

    if (code) {
      const { client_id, token_endpoint } = config[provider];
      const code_verifier = window.sessionStorage.getItem("code_verifier");
      const payload = {
        grant_type: "authorization_code",
        code,
        token_endpoint,
        client_id,
        code_verifier,
      };
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });
      if (!res.ok) {
        setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
        return;
      }
      const { access_token } = await res.json();
      await registerLogin(provider, { access_token });
      reload();
      navigate("/");
      return;
    }

    if (!access_token) {
      setError("Missing access_token");
      return;
    }

    await registerLogin(provider, { access_token });
    reload();
    navigate("/");
  }, []);

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return <h1>Please wait...</h1>;
}

export function EndSession({ reload }) {
  const navigate = useNavigate();
  const { endSession } = useContext(ApiContext);
  useEffect(async () => {
    await endSession();
    reload();
    navigate("/");
  });
  return <h1>Please wait...</h1>;
}

function LoginButton({ config, label, provider }) {
  async function handleLogin() {
    const { authorization_endpoint, response_type, scope, client_id } =
      config[provider];

    const state = randomString(50);
    window.sessionStorage.setItem("expected_state", state);

    const parameters = {
      response_type,
      response_mode: "fragment",
      client_id,
      state,
      scope,
      redirect_uri: `${window.location.origin}/login/${provider}/callback`,
    };

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(parameters);
  }

  return (
    <div>
      <img src={Google} className={"google-icon"}/>
      <button className={"google-login-btn"} onClick={handleLogin}>{label}</button>
      <h2 className={"line"}>________________________</h2>
    </div>
  );
}

function StartLogin({ config }) {
  return (
    <div>
      <div className={"login-container"}>
        <div className={"sidebar"}>
          <img src={FirstLayer} className={"first-layer-img"}/>
          <img src={SecondLayer} className={"second-layer-img"}/>
          <img src={TopCorner} className={"top-corner-img"}/>
          <img src={BottomCorner} className={"bottom-corner-img"}/>
        </div>
        <div className={"login-form"}>
          <h1>Get started.</h1>
          <form>
            <LoginButton
              label={"Sign up with Google"}
              config={config}
              provider={"google"}
            />
            <div className={"email-field"}>
              <label className={"login-txt"}>Email address</label>
              <img src={Message} className={"message-icon"}/>
              <input type={"email"} value={"@example.com"} className={"login-txt"}/>
            </div>
            <div>
              <label className={"login-txt"}>Password</label>
              <img src={Lock} className={"lock-icon"}/>
              <img src={Invisible} className={"invisible-icon"}/>
              <input type={"password"} value={"exmplepassword"} className={"login-txt"}/>
            </div>
            <div>
              <label><input type={"checkbox"} className={"checkbox"}/>I agree to Meliora Impacts’s Terms of Service and Privacy Policy</label>
            </div>
           <button className={"sign-in-btn"}>Sign in</button>
            <label className={"sign-up-txt"}>Don´t have an account? Sign up</label>
          </form>
        </div>
      </div>
    </div>
  );
}

export function LoginPage({ config, reload }) {
  return (
    <Routes>
      <Route path={"/"} element={<StartLogin config={config} />} />
      <Route
        path={"/:provider/callback"}
        element={<LoginCallback config={config} reload={reload} />}
      />
      <Route path={"/endsession"} element={<EndSession reload={reload} />} />
      <Route path={"*"} element={<StartLogin config={config} />} />
    </Routes>
  );
}