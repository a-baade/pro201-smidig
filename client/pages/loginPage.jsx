import React, { useContext, useEffect, useState } from "react";
import { Link, Route, Routes, useNavigate, useParams } from "react-router-dom";
import { ApiContext } from "../apiContext";
import { randomString } from "../lib/randomString";
import Google from "../pics/sign-up-pic/img.png";
import Message from "../pics/sign-up-pic/img_2.png";
import Lock from "../pics/sign-up-pic/img_3.png";
import { SignUp } from "./signUp";
import FirstLayer from "../pics/sign-in/firstLayer.png";
import SecondLayer from "../pics/sign-in/secondLayeer.png";
import ThirdLayer from "../pics/sign-in/thirdLayer.png";
import LoadingWheel from "../components/loadingSpinner";

export function LoginCallback({ reload, config }) {
  const { provider } = useParams();
  const [error, setError] = useState();
  const navigate = useNavigate();
  const { registerLogin } = useContext(ApiContext);
  useEffect(async () => {
    const { access_token, error, error_description, state, code } =
      Object.fromEntries(
        new URLSearchParams(window.location.hash.substring(1))
      );

    const expected_state = window.sessionStorage.getItem("expected_state");
    if (!state || expected_state !== state) {
      setError("Unexpected state");
      return;
    }

    if (error || error_description) {
      setError(`Error: ${error} (${error_description})`);
      return;
    }

    if (code) {
      const { client_id, token_endpoint } = config[provider];
      const code_verifier = window.sessionStorage.getItem("code_verifier");
      const payload = {
        grant_type: "authorization_code",
        code,
        token_endpoint,
        client_id,
        code_verifier,
      };
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });
      if (!res.ok) {
        setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
        return;
      }
      const { access_token } = await res.json();
      await registerLogin(provider, { access_token });
      reload();
      navigate("/");
      return;
    }

    if (!access_token) {
      setError("Missing access_token");
      return;
    }

    await registerLogin(provider, { access_token });
    reload();
    navigate("/");
  }, []);

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error.toString()}</div>
      </div>
    );
  }

  return <LoadingWheel/>;
}

export function EndSession({ reload }) {
  const navigate = useNavigate();
  const { endSession } = useContext(ApiContext);
  useEffect(async () => {
    await endSession();
    reload();
    navigate("/");
  });
  return <LoadingWheel/>;
}

export function LoginButton({ config, label, provider }) {
  async function handleLogin() {
    const { authorization_endpoint, response_type, scope, client_id } =
      config[provider];

    const state = randomString(50);
    window.sessionStorage.setItem("expected_state", state);

    const parameters = {
      response_type,
      response_mode: "fragment",
      client_id,
      state,
      scope,
      redirect_uri: `${window.location.origin}/login/${provider}/callback`,
    };

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(parameters);
  }

  return (
    <div>
      <img src={Google} className={"google-icon"} alt={"googleIcon"} />
      <button className={"google-login-btn"} onClick={handleLogin}>
        {label}
      </button>
      <h2 className={"line"}></h2>
    </div>
  );
}

function StartLogin({ config }) {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState();
  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();
    const result = await fetch("/api/login", {
      method: "POST",
      body: JSON.stringify({ email, password }),
      headers: {
        "content-type": "application/json",
      },
    });
    if (result.ok) {
      navigate("/");
    } else {
      setError("Login failed");
    }
  }

  return (
    <div className={"page-margin"}>
      <div className={"login-container"}>
        <div className={"sidebar"}>
          <img src={FirstLayer} className={"first-layer-login"} />
          <img src={SecondLayer} className={"second-layer-login"} />
          <img src={ThirdLayer} className={"third-layer-login"} />
        </div>
        <div className={"login-form"}>
          <h1>The future is NOW.</h1>

          <LoginButton
            label={"Sign in with Google"}
            config={config}
            provider={"google"}
          />
          <form onSubmit={handleSubmit}>
            <div className={"email-field"}>
              <label className={"login-txt"}>Email address</label>
              <img src={Message} className={"message-icon"} alt={"msgIcon"} />
              <input
                type="email"
                value={email}
                onChange={(e) => setEmail(e.target.value)}
                placeholder={"email@address.com"}
                className={"login-txt"}
                name={"email"}
              />
            </div>
            <div>
              <label className={"login-txt"}>Password</label>
              <img src={Lock} className={"lock-icon"} alt={"lockIcon"} />
              <input
                type={"password"}
                value={password}
                onChange={(e) => setPassword(e.target.value)}
                className={"login-txt"}
                placeholder={"*************"}
                name={"password"}
              />
            </div>
            <div>{error && <div>{error}</div>}</div>
            <button className={"sign-in-btn"}>Sign in</button>
            <label className={"sign-up-txt"}>
              DonÂ´t have an account? <Link to={"/signup"}>Sign up</Link>{" "}
            </label>
          </form>
        </div>
      </div>
    </div>
  );
}

export function LoginPage({ config, reload }) {
  return (
    <Routes>
      <Route path={"/"} element={<StartLogin config={config} />} />
      <Route path={"/signup"} element={<SignUp config={config} />} />
      <Route
        path={"/:provider/callback"}
        element={<LoginCallback config={config} reload={reload} />}
      />
      <Route path={"/endsession"} element={<EndSession reload={reload} />} />
      <Route path={"*"} element={<StartLogin config={config} />} />
    </Routes>
  );
}
