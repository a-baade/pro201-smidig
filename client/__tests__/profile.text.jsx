import { Profile } from "../profile";
import ReactDOM from "react-dom";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { createRoot } from "react-dom/client";
import { ProfileContext } from "../loginPage";
import { act } from "react-dom/test-utils";

describe("Profile component", () => {
  it("should test div element", async function () {
    const divEl = document.createElement("div");
    const root = createRoot(divEl);
    root.render(
      <MemoryRouter>
        <Profile />
      </MemoryRouter>
    );
    expect(divEl.getElementsByTagName("div")).toMatchSnapshot();
  });

  it("should test text in h1 element", function () {
    const h1El = document.createTextNode(/Profile for/i);
    expect(h1El).toMatchSnapshot();
  });

  it("should show demo profile", async function () {
    const userinfo = {
      userinfo: { name: "test name", email: "test email" },
    };
    const divEl = document.createElement("div");
    const root = createRoot(divEl);
    await act(async () => {
      root.render(
        <ProfileContext.Provider value={{ userinfo: () => userinfo }}>
          <Profile />
        </ProfileContext.Provider>
      );
    });
    expect(
      Object.values(divEl.querySelectorAll("h1")).map((e) => e.innerHTML)
    ).toEqual(["Profile for userinfo ()"]);
    expect(divEl.innerHTML).toMatchSnapshot();
    expect.stringContaining("Profile for test name (test email)");
  });
});
