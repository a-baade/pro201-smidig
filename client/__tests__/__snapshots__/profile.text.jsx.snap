// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component should show demo profile 1`] = `"<div><h1>Profile</h1><div><div><h3>Name: test name</h3><h3>Email: test email</h3></div><div><img alt=\\"Profile picture\\"></div></div></div>"`;

exports[`Profile component should test div element 1`] = `
HTMLCollection [
  <div>
    <h1>
      Profile
    </h1>
    <div>
      <div>
        <h3>
          Name: 
          test name
        </h3>
        <h3>
          Email: 
          test email
        </h3>
      </div>
      <div>
        <img
          alt="Profile picture"
        />
      </div>
    </div>
  </div>,
  <div>
    <div>
      <h3>
        Name: 
        test name
      </h3>
      <h3>
        Email: 
        test email
      </h3>
    </div>
    <div>
      <img
        alt="Profile picture"
      />
    </div>
  </div>,
  <div>
    <h3>
      Name: 
      test name
    </h3>
    <h3>
      Email: 
      test email
    </h3>
  </div>,
  <div>
    <img
      alt="Profile picture"
    />
  </div>,
]
`;

exports[`Profile component should test text in h1 element 1`] = `/Profile for/i`;
